readme.md
# FibX

***

**fibx** 是 fibjs 的一个 web 框架,提供了中间件安装以及请求接受和应答的功能

## Installation

```
npm install @fibjs/fibx
```

## Example     

```javascript
var app = require('@fibjs/fibx')();
var http = require('http');

app.use(function(next) {

    this.cookies.set('hello', 'me');
    this.state = {};
    this.state.number = 1;
    this.state['back'] = 'fibx';
    next();
    this.body = 'hello world! ------' + this.state['back'] + 'number:' + this.state.number;
});

//设置1000个中间件
for (var i = 0; i < 1000; i++) {
    app.use(function(next) {
        this.state.number++;
        next && next();
    });
}

//app.use 的用法详见 Api Doc
app.use('^/go(/.*)$', http.fileHandler('./'));

app.use('^/a(/.*)$', function(next) {
    this.state.number = -1;
    next();
});
app.use('^/a(/.*)$', function() {
    this.state.number = -2;
});

app.use('^(/.*)$', function() {
    this.state.number++;
});

app.listen(10023);
```

## Api

[API Doc](https://github.com/fibx/fibx/blob/master/doc/api.md)

## Next      				

more 中间件 : 					

* session										
* static			
* body parse			
* and more ...


CHANGELOG.md
* v0.4.0

```
1. 支持 fibx.use 直接 native middleware
2. 支持 fibjs 0.8.0 版本
3. 修复一些测试用例等兼容问题
```

* v0.5.0

```
1. 支持 fibjs 0.20.0 版本
2. 修复一些测试用例等兼容问题
```

test.js
/**
 * @author Rube
 * @date 15/12/15
 */

var app = require('./index')();
var http = require('http');

app.use(function(next) {

    this.cookies.set('hello', 'me');
    this.state = {};
    this.state.number = 1;
    this.state['back'] = 'fibx';
    next();
    this.body = 'hello world! ------' + this.state['back'] + 'number:' + this.state.number;
});

for (var i = 0; i < 1000; i++) {
    app.use(function(next) {
        this.state.number++;
        next && next();
    });
}

app.use('^/go(/.*)$', http.fileHandler('./'));

app.use('^/a(/.*)$', function(next) {
    this.state.number = -1;
    next();
});
app.use('^/a(/.*)$', function() {
    this.state.number = -2;
});

app.use('^(/.*)$', function() {
    this.state.number++;
});

app.listen(10023);


test/index.js
var test = require('test');
var http = require('http');
var coroutine = require('coroutine');
var fs = require('fs');
var md5 = require('hash').md5;
var app = require('../index')();

coroutine.start(function() {

    /** fibx request **/
    app.use('^/request/basic$', function() {
        this.body = {
            method: this.method,
            keepAlive: this.keepAlive,
            path: this.path,
            protocol: this.protocol,
            ip: this.ip,
            port: this.port,
            length: this.length,
            type: this.type,
            host: this.host
        };
    });

    app.use('^/request/query$', function() {
        this.body = this.query;
    });

    app.use('^/request/form$', function() {
        this.body = this.form;
    });

    app.use('^/request/type$', function() {
        this.body = this.is('^text/xm[a-z]+$');
    });

    /** fibx cookies **/
    app.use('^/cookies/signed$', function() {
        if (this.query.c === 'get') {
            console.log(this.r.cookies.all('b'));
            console.log(this.r.cookies.all('b.sig'));
            this.body = this.cookies.get('b');
        }
        if (this.query.c === 'set') {
            this.cookies.set('b', 'papapa');
            this.body = 'papapa';
        }
    });

    app.use('^/cookies/nosigned$', function() {
        if (this.query.c === 'get') {
            this.body = this.cookies.get('a', {signed: false});
        }
        if (this.query.c === 'set') {
            this.cookies.set('a', 'papapa', {signed: false});
            this.body = 'papapa';
        }
    });


    /** fibx response **/
    app.use('^/response/number$', function() {
        this.body = 1;
    });

    app.use('^/response/string$', function() {
        this.body = 'hello';
    });

    app.use('^/response/object2json$', function() {
        this.body = {a: 1};
    });

    app.use('^/response/boolean$', function() {
        this.body = false;
    });

    app.use('^/response/stream$', function() {
        this.body = fs.openFile('./index.js');
    });

    app.use('^/response/status$', function() {
        this.status = 500;
        this.type = 'text/xml';
    });

    app.use('^/response/redirect$', function() {
        this.redirect('http://localhost:5210/response/number');
    });

    /** fibx middleware **/
    app.use('^/next$', function(next) {

        this.state = {};
        this.state.number = 1;
        next && next();
        this.body = 'number:' + this.state.number;
    });
    app.use('^/next$', function() {
        this.state.number++;
    });

    app.use('^/nextr$', function(next) {

        this.state = {};
        this.state.number = 1;
        this.body = next();
    });
    app.use('^/nextr$', function() {
        this.state.number += 10;
        return 'number:' + this.state.number;
    });

    app.use('^/nextp$', function(next) {

        this.state = {};
        this.body = next('hello Rube');
    });
    app.use('^/nextp$', function(value) {
        return value;
    });
    /** fibx basic **/
    app.use(function(next) {
        this.state = {body: 'hello world'};
        next && next();
    });
    app.use('^/fibx$', function() {
        this.body = 'hello a';
    });
    app.use('^/fileHandle/(.*)$', http.fileHandler('./'));
    app.use('/', function() {
        this.body = 'hello b';
    });
    app.use('/__all__', function(){
        this.body = this.state.body;
    });
    app.listen(5210);
});
coroutine.sleep(100);
test.setup();

describe('-----------------------fibx----------------------\r\n', function() {

    describe('fibx basic', function() {

        it('fibx start and listen the port', function() {
            var r = http.request('get', 'http://127.0.0.1:5210/');
            assert.equal(r.read().toString(), 'hello b');
        });

        it('fibx __all__ middleware execute', function() {
            var r = http.request('get', 'http://127.0.0.1:5210/__all__');
            assert.equal(r.read().toString(), 'hello world');
        });

        it('fibx route middleware execute', function() {
            var r = http.request('get', 'http://127.0.0.1:5210/fibx');
            assert.equal(r.read().toString(), 'hello a');
        });

        it('fibx fileHandler can be used', function() {
            var r = http.request('get', 'http://127.0.0.1:5210/fileHandle/index.js');
            var text = fs.readFile('./index.js').toString();
            assert.equal(text, r.read().toString());
        });

        it('request is independent', function() {
            var r = http.request('get', 'http://127.0.0.1:5210/response/status');
            assert.equal(r.headers.first('Content-Type'), 'text/xml');
            assert.equal(r.statusCode, 500);

            var r = http.request('get', 'http://127.0.0.1:5210/response/number');
            assert.equal(r.statusCode, 200);
        });
    });

    describe('fibx middleware', function() {

        it('multi middlewares', function() {
            var r = http.request('get', 'http://127.0.0.1:5210/next');
            assert.equal(r.read().toString(), 'number:2');
        });

        it('fibx can receive post,get,and more', function() {
            var r1 = http.request('post', 'http://127.0.0.1:5210/next');
            var r2 = http.request('delete', 'http://127.0.0.1:5210/next');
            var r3 = http.request('put', 'http://127.0.0.1:5210/next');
            var r4 = http.request('get', 'http://127.0.0.1:5210/next');
            assert.equal(r1.read().toString(), 'number:2');
            assert.equal(r2.read().toString(), 'number:2');
            assert.equal(r3.read().toString(), 'number:2');
            assert.equal(r4.read().toString(), 'number:2');
        });

        it('next return value', function() {
            var r = http.request('get', 'http://127.0.0.1:5210/nextr');
            assert.equal(r.read().toString(), 'number:11');
        });

        it('next pass value', function() {
            var r = http.request('get', 'http://127.0.0.1:5210/nextp');
            assert.equal(r.read().toString(), 'hello Rube');
        });
    });

    describe('fibx request', function() {

        it('request basic', function() {
            var r = http.request('get', 'http://127.0.0.1:5210/request/basic');
            var res = {
                "method": "get",
                "keepAlive": true,
                "path": "/request/basic",
                "protocol": "HTTP/1.1",
                "ip": "127.0.0.1",
                "length": 0,
                "host": "127.0.0.1:5210"
            };
            var sure = JSON.parse(r.read().toString());
            assert.equal(typeof sure.port, 'number');
            delete sure.port;
            assert.equal(JSON.stringify(res), JSON.stringify(sure));
        });

        it('request query', function() {
            var r = http.request('get', 'http://127.0.0.1:5210/request/query?hello=1&world=2');
            assert.equal(r.read().toString(), JSON.stringify({hello: "1", world: "2"}));
        });

        it('request form', function() {
            var r = http.post('http://127.0.0.1:5210/request/form', {body: "hello=a&world=b", headers: {"Content-Type": "application/x-www-form-urlencoded"}});
            assert.equal(r.read().toString(), JSON.stringify({hello: "a", world: "b"}));
        });

        it('request type is/get', function() {
            var r = http.request('get', 'http://127.0.0.1:5210/request/type', {headers: {"Content-Type": "text/xml"}});
            var r1 = http.request('get', 'http://127.0.0.1:5210/request/type',{headers: {"Content-Type": "text/xm1"}});
            assert.equal('false', r1.read().toString());
            assert.equal('true', r.read().toString());
        });
    });

    describe('fibx response', function() {

        function handler(resData, type, path) {
            app.handler({
                response: {
                    write: function(data) {
                        assert.equal(data, resData);
                    },
                    headers: {
                        set: function(k, v) {
                            if (typeof type === 'string') {
                                assert.equal(v, type);
                            } else {
                                assert.ok((v === type[0] || v === type[1]));
                            }
                        }
                    }
                },
                stream: {},
                headers: {}
            }, path);
        }

        it('response number', function() {
            handler(1, 'text/html', '^/response/number$');
        });

        it('response boolean', function() {
            handler('false', 'text/html', '^/response/boolean$');
        });

        it('response string', function() {
            handler('hello', 'text/html', '^/response/string$');
        });

        it('response object2json', function() {
            handler(JSON.stringify({a: 1}), ['text/html', 'text/json'], '^/response/object2json$');
        });

        it('response stream', function() {
            var r = http.request('get', 'http://127.0.0.1:5210/response/stream');
            assert.equal(r.read().toString(), fs.readFile('./index.js', 'utf8').toString());
        });

        it('response redirect', function() {
            var r = http.request('get', 'http://127.0.0.1:5210/response/redirect');
            assert.equal(r.read().toString(), 1);
        });

        it('response status and content-type', function() {
            var r = http.request('get', 'http://127.0.0.1:5210/response/status');
            assert.equal(r.headers.first('Content-Type'), 'text/xml');
            assert.equal(r.statusCode, 500);
        });
    });

    describe('fibx cookies', function() {

        it('set signed cookie', function() {
            var value = 'Rube';
            var value_hex =md5(md5('bRube').digest().hex() + app.key).digest().hex();
            var r = http.get('http://127.0.0.1:5210/cookies/signed?c=get',{headers: {"cookie":"b=" + value + "; path=/; b.sig=" + value_hex + "; path=/;"}});
            assert.equal(value, r.read().toString());
        });

        it('get signed cookie', function() {
            var r = http.get('http://127.0.0.1:5210/cookies/signed?c=set');
            assert.equal(r.cookies[0].value, r.read().toString());
        });

        it('set nosigned cookie', function() {
            var value = 'Rube';
            var r = http.get('http://127.0.0.1:5210/cookies/nosigned?c=get',{headers :{"cookie":"a=" + value + "; path=/;"}});
            assert.equal(value, r.read().toString());
        });

        it('get nosigned cookie', function() {
            var r = http.get('http://127.0.0.1:5210/cookies/nosigned?c=set');
            assert.equal(r.cookies[0].value, r.read().toString());
        });
    });
});

test.run();
process.exit(0);


test/simple-example.js
/**
 * @author Rube
 * @date 15/12/15
 */

var app = require('../index')();
var http = require('http');
var fs = require('fs');

app.use(function(next) {

    this.cookies.set('hello', 'me');
    this.state = {};
    this.state.number = 1;
    this.state['back'] = 'fibx';
    var a = next();
    if (a) {
        this.body = a + this.state.number;
    }
});

for (var i = 0; i < 1000; i++) {
    app.use(function(next) {
        this.state.number++;
        return next && next();
    });
}

app.use('^/request/basic$', function() {
    this.body = {
        method: this.method,
        keepAlive: this.keepAlive,
        path: this.path,
        protocol: this.protocol,
        ip: this.ip,
        port: this.port,
        length: this.length,
        type: this.type,
        host: this.host
    };
});

app.use('^/all(/.*)$', function() {
    this.state.number = 'all';
    return 'hello world Rube~';
});

app.use('^/s.zip$', function() {
    var stream = fs.open(__dirname + '/simple-example.js');
    this.type = 'application/octet-stream';
    this.body = stream;
});

app.use('^/go(/.*)$', http.fileHandler('./'));

app.use('^/a(/.*)$', function(next) {
    this.state.number = -1;
    next();
});
app.use('^/a(/.*)$', function() {
    this.state.number = -2;
});

app.use('^(/.*)$', function() {
    this.state.number++;
});

app.listen(10023);


index.js
/**
 * @author Rube
 * @date 15/12/15
 */

var http = require('http');
var mq = require('mq');
var connect = require('./lib/connect');
var request = require('./lib/request');
var response = require('./lib/response');
var cookies = require('./lib/cookies');

var app = Application.prototype;
var methodCollection = {};

module.exports = Application;

function Application() {
    if (!(this instanceof Application)) return new Application();
    this.middlewares = {'__all__': []};
    this.nativeMiddlewares = [];
    this.connected = {};
    this.routing = {};
    request.init.call(methodCollection);
    cookies.init.call(methodCollection);
    response.init.call(methodCollection);
}

app.listen = function(port) {
    var that = this;
    for (var route in this.middlewares) {
        (function(route) {
            that.connected[route] = function(cxt) {
                return connect.call(cxt, that.middlewares['__all__'].concat(that.middlewares[route]));
            };
            that.routing[route] = function(r) {
                that.handler.call(that, r, route);
            }
        })(route);
    }

    var svr = new http.Server(port, this.nativeMiddlewares.concat(new mq.Routing(that.routing)));
    svr.run();
};

app.use = function(middleware) {
    if (middleware.toString().toLocaleLowerCase() === 'handler'){
        this.nativeMiddlewares.push(middleware);
        return;
    }

    if (typeof middleware === 'function') {
        this.middlewares['__all__'].push(middleware);
    } else {
        var argv = Array.prototype.slice.apply(arguments);
        if (argv[1].toString().toLocaleLowerCase() === 'handler') {
            this.routing[argv[0]] = argv[1];
            return;
        }
        this.middlewares[argv[0]] = this.middlewares[argv[0]] || [];
        this.middlewares[argv[0]].push(argv[1]);
    }
};

app.handler = function(r, route) {
    var _r = Object.create({});
    Object.assign(_r, methodCollection);
    _r.r = r;
    _r.key = app.key;
    request.run.call(_r, r);
    cookies.run.call(_r, r);
    this.connected[route](_r).call(_r);
    response.run.call(_r, r);
};

app.key = 'fibx';

lib/cookies.js
/**
 * @author Rube
 * @date 15/12/16
 */
var http = require('http');
var md5 = require('hash').md5;

var cookies = {
    cookies: {
        get: function(name, option) {
            if (option && option.signed === false) {
                return getCookies.call(this, name);
            } else {
                return getSignedCookies.call(this, name);
            }
        },
        set: function(name, value, options) {
            var argv = Array.prototype.slice.call(arguments);
            if (options && options.signed === false) {
                setCookies.apply(this, argv);
            } else {
                setSignCookies.apply(this, argv);
            }
        }
    }
};

exports.init = function() {
    Object.assign(this, cookies);
};

exports.run = function(r) {
    this.cookies['r'] = r;
    this.cookies['key'] = this.key;
};

function setSignCookies(name, value, options) {
    var signValue = md5(md5(name + value).digest().hex() + this.key).digest().hex();
    setCookies.call(this, name, value, options);
    setCookies.call(this, name + '.sig', signValue, options);
};

function getSignedCookies(name) {

    var value = this.r.cookies[name];
    var signValue = this.r.cookies[name + '.sig'];
    if (signValue && value && md5(md5(name + value).digest().hex() + this.key).digest().hex() == signValue) {
        return value;
    } else {
        return null;
    }
};

function setCookies(name, value, options) {
    options = options || {};
    options['path'] = options['path'] || '/';
    options['httpOnly'] = options['httpOnly'] || true;
    options['expires'] = options['expires'] || new Date();
    this.r.response.addCookie(new http.Cookie(name, value, options));
};

function getCookies(name) {
    if (!!!this.r.cookies[name + '.sig']) {
        return this.r.cookies[name];
    } else {
        return null;
    }
};


lib/request.js
/**
 * @author Rube
 * @date 15/12/16
 */

var request = {
    get: function(field) {
        return this.r.headers[field];
    },
    is: function(type) {
        var reg = new RegExp(type);
        return reg.test(this.get('Content-Type'));
    }
};

exports.init = function() {
    Object.assign(this, request);
};

exports.run = function(r) {
    this.request = {};
    this.method = r.method;
    this.keepAlive = r.keepAlive;
    this.query = {};
    this.form = {};
    this.queryString = r.queryString;
    this.path = r.address;
    this.header = r.headers;
    this.protocol = r.protocol;
    if (r.socket) {
        this.ip = r.socket.remoteAddress;
        this.port = r.socket.remotePort;
    }
    this.length = r.length;
    this.type = this.get('Content-Type');
    this.host = this.get('Host');
    for (var key in r.query) {
        if (!(typeof r.query[key] == 'function')) {
            this.query[key] = r.query[key];
        }
    }
    for (var key in r.form) {
        if (!(typeof r.form[key] == 'function')) {
            this.form[key] = r.form[key];
        }
    }
};

lib/response.js
/**
 * @module
 * @author Rube
 * @date 15/12/16
 * @desc
 */

var response = {
    set: function(field, value) {
        if (typeof field == 'string') {
            this.r.response.headers.set(field, value);
        } else {
            for (var f in field) {
                this.r.response.headers.set(f, field[f]);
            }
        }
    },
    remove: function(field) {
        this.r.response.headers.remove(field);
    },
    redirect: function(url) {
        this.status = 302;
        this.r.response.redirect(url);
    }
};

exports.init = function() {
    Object.assign(this, response);
};

exports.run = function(r) {

    if (this.body === false) {
        this.body = 'false'
    }
    this.body = this.body || '';
    this.status = this.status || 200;
    this.type = this.type || '';

    var res = r.response;
    res.statusCode = this.status;

    if (this.type) {
        this.set('Content-Type', this.type);
    } else {
        this.set('Content-Type', 'text/html');
    }

    if (this.body.toString().toLowerCase().indexOf('stream') != -1 ||
        this.body.toString().toLowerCase().indexOf('file') != -1) {

        res.body = this.body;
        return;
    }

    switch (typeof this.body) {
        case 'number':
            this.body += '';
            break;
        case 'object':
            this.body = JSON.stringify(this.body);
            this.set('Content-Type', 'text/json');
            break;
        case 'boolean':
            this.body += '';
            break;
    }

    res.write(this.body);
};